runtime/src/stakes.rs

```diff
@@ -225,24 +225,58 @@ impl Stakes<StakeAccount> {
     /// full account state for respective stake pubkeys. get_account function
     /// should return the account at the respective slot where stakes where
     /// cached.
-    pub(crate) fn new<F>(stakes: &Stakes<Delegation>, get_account: F) -> Result<Self, Error>
+    pub(crate) fn new<F>(stakes: Stakes<Delegation>, get_account: F) -> Result<Self, Error>
     where
-        F: Fn(&Pubkey) -> Option<AccountSharedData>,
+        F: Fn(&Pubkey) -> Option<AccountSharedData> + Sync,
     {
-        let stake_delegations = stakes.stake_delegations.iter().map(|(pubkey, delegation)| {
-            let Some(stake_account) = get_account(pubkey) else {
-                return Err(Error::StakeAccountNotFound(*pubkey));
-            };
-            let stake_account = StakeAccount::try_from(stake_account)?;
-            // Sanity check that the delegation is consistent with what is
-            // stored in the account.
-            if stake_account.delegation() == *delegation {
-                Ok((*pubkey, stake_account))
-            } else {
-                Err(Error::InvalidDelegation(*pubkey))
-            }
-        });
+        // im::HashMap doesn't support rayon so we manually build a temporary vector. Note this is
+        // what std HashMap::par_iter() does internally too.
+        let stake_delegations_vec = stakes.stake_delegations.into_iter().collect::<Vec<_>>();
+        let stake_delegations = stake_delegations_vec
+            .into_par_iter()
+            // We use fold/reduce to aggregate the results, which does a bit more work than calling
+            // collect_vec_list() and then manually inserting into a map, but it does it in background
+            // threads, so effectively it's faster.
+            .try_fold_with(ImHashMap::new(), |mut map, (pubkey, delegation)| {
+                let Some(stake_account) = get_account(&pubkey) else {
+                    return Err(Error::StakeAccountNotFound(pubkey));
+                };
+
+                // Assert that all valid vote-accounts referenced in stake delegations are already
+                // contained in `stakes.vote_account`.
+                let voter_pubkey = &delegation.voter_pubkey;
+                if stakes.vote_accounts.get(voter_pubkey).is_none() {
+                    if let Some(account) = get_account(voter_pubkey) {
```

Comment:
Note that I have removed the HashSet in order to parallelize. The rationale is that in the common case - not a malicious or somehow broken snapshot  - is_none() is never true and we never go into the if, so we have optimal perf. With a malicious snapshot we do a constant (stake_delegations.len()) amount of extra work, which is not an attack vector.

Reply:
What happens if there are a lot of stake accounts delegated to a specific vote account and then the vote account is deleted or something (e.g. the validator node shuts down)?
Also not spilling the details here, but we need to think if this can be abused in malicious ways (which I am suspecting it can).

Reply:
The same that was happening before? I don't think my PR changes anything related to this?

Reply:
isn't the old only checking unique voter-pubkeys but the new code doing it once for each stake-account delegated to that voter-pubkey?

Reply:
Correct, hence

>  With a malicious snapshot we do a constant (stake_delegations.len()) amount of extra work, which is not an attack vector.

am I wrong here?

Reply:
Probably we shouldn't try to find attack vectors in the open.
But do you agree in this case
> What happens if there are a lot of stake accounts delegated to a specific vote account and then the vote account is deleted or something (e.g. the validator node shuts down operations and leaves)?

the new code does more work ?

Reply:
> Probably we shouldn't try to find attack vectors in the open.

The code is not merged, so why not? Or are you saying that the existing code might be vulnerable, in which case we can address that in a separate PR? 😋

> But do you agree in this case
>>    What happens if there are a lot of stake accounts delegated to a specific vote account and then the vote account is deleted or something (e.g. the validator node shuts down operations and leaves)?

> the new code does more work ?

In the common case - where a non malicious validator has produced a snapshot - no it does not do more work than what's in master right now. In that case we do `O(get_account) * stakes.stake_delegations.len()`. This is because in the common case we never go inside this if:

```
                // Assert that all valid vote-accounts referenced in stake delegations are already
                // contained in `stakes.vote_account`.
                let voter_pubkey = &delegation.voter_pubkey;
                if stakes.vote_accounts.get(voter_pubkey).is_none() {
                    if let Some(account) = get_account(voter_pubkey) {
                        if VoteStateVersions::is_correct_size_and_initialized(account.data())
                            && VoteAccount::try_from(account.clone()).is_ok()
                        {
                            error!("vote account not cached: {voter_pubkey}, {account:?}");
                            return Err(Error::VoteAccountNotCached(*voter_pubkey));
                        }
                    }
                }
```

`stakes.vote_accounts` always contains `voter_pubkey` - it's the thing we're asserting.

With a malicious snapshot that triggers the worst case, we go inside the if at every iteration. We can have maximum `stakes.stake_delegations.len()` iterations, so we do `O(get_account) * stakes.stake_delegations.len() * 2` - we do a get_account() outside the if and one inside the if.

Since in the worst case we do a constant amount of extra work, for a very small constant, I don't think this changes anything security wise.